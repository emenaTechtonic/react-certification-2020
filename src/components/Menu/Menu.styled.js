import styled from 'styled-components'

const StyledMenu = styled.div.attrs({
  className: 'menu',
})`
  display: flex;

  flex-direction: column;
  justify-content: flex-start;
  background: ${({ toggle }) => (toggle ? '#2d314d' : '#FFFFFF')};
  height: 100vh;
  text-align: left;
  width: ${({ toggle }) => (toggle ? '10em' : '0em')};
  height: 100vh;
  position: fixed;
  top: 0;
  left: 0;
  transition-property: background, width;
  transition-duration: 0.3s, 0.3s;
  transition-timing-function: ease-in, ease-in;
  a {
    margin-left: 1em;
    margin-top: 2em;
    font-weight: bold;
    opacity: ${({ toggle }) => (toggle ? '1' : '0')};
    transition: ${({ toggle }) =>
      toggle ? 'opacity 0.7s ease-in' : 'opacity 0.2s ease-in'};
    &:last-child {
      margin-top: 0em;
    }
  }
`
export default StyledMenu
