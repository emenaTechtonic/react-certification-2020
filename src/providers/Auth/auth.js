import React, { useState, useEffect, createContext, useCallback } from 'react'
import { AUTH_STORAGE_KEY, AUTH_STORAGE_USER } from '../../utils/constants'
import { storage } from '../../utils/storage'

export const AuthContext = createContext(null)

export const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false)

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY)
    const isAuthenticated = Boolean(lastAuthState)

    setAuthenticated(isAuthenticated)
  }, [])

  const login = useCallback((user) => {
    setAuthenticated(true)
    storage.set(AUTH_STORAGE_KEY, true)
    storage.set(AUTH_STORAGE_USER, user)
  }, [])

  const logout = useCallback((user) => {
    setAuthenticated(false)
    storage.set(AUTH_STORAGE_KEY, false)
    storage.set(AUTH_STORAGE_USER, user)
  }, [])

  return (
    <AuthContext.Provider value={{ login, logout, authenticated }}>
      {children}
    </AuthContext.Provider>
  )
}

// export { useAuth }
export default AuthProvider
