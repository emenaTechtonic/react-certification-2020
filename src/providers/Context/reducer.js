export const ContextReducer = (state, action) => {
  switch (action.type) {
    case 'TOGGLE_THEME':
      return {
        ...state,
        theme: !state.theme,
      }
    case 'OPEN_MENU':
      return {
        ...state,
        menu: action.payload,
      }
    case 'GET_DATA':
      return {
        ...state,
        data: action.payload,
      }
    case 'SET_VIEW':
      return {
        ...state,
        view: action.payload,
      }
    case 'SET_TARGET':
      return {
        ...state,
        target: action.payload,
      }
    case 'SET_USER':
      return {
        ...state,
        user: action.payload,
      }
    case 'SET_FAVORITES':
      return {
        ...state,
        favorites: action.payload,
      }

    default:
      return state
  }
}
