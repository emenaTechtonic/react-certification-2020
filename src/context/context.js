import React, { useEffect, createContext, useReducer, memo } from 'react'
import { useFetch } from '../utils/hooks/useFetch'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data, isLoading } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])

  return isLoading ? (
    <div> Loading...</div>
  ) : (
    <Context.Provider value={{ state, dispatch }}> {children} </Context.Provider>
  )
}
export default memo(ContextProvider)
