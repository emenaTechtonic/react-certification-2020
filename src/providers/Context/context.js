import React, { useEffect, createContext, useCallback, useReducer, memo } from 'react'
import { useFetch } from '../../utils/hooks/useFetch'
import { storage } from '../../utils/storage'
import { isElementIn, removeElement } from '../../utils/fns'
import {
  AUTH_STORAGE_USER,
  AUTH_STORAGE_KEY,
  AUTH_STORAGE_FAVORITES,
} from '../../utils/constants'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
  user: { id: 0, name: '', avatarUrl: '' },
  favorites: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])
  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY)
    const isAuthenticated = Boolean(lastAuthState)

    if (isAuthenticated) {
      const lastUser = storage.get(AUTH_STORAGE_USER)
      const userAuthenticated = Object(lastUser)

      dispatch({ type: 'SET_USER', payload: userAuthenticated })

      const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
      if (lastFavorites) {
        const favorites = lastFavorites
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
      }
    }
  }, [])

  const saveFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const res = isElementIn(card.id, favorites)
      if (!res) {
        favorites.push(card)
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
        storage.set(AUTH_STORAGE_FAVORITES, favorites)
      }
    }
  }, [])
  const removeFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const newFavorites = removeElement(card.id, favorites)
      dispatch({ type: 'SET_FAVORITES', payload: newFavorites })
      storage.set(AUTH_STORAGE_FAVORITES, newFavorites)
    }
  }, [])

  // return isLoading ? (
  //   <div> Loading...</div>
  // ) : (
  return (
    <Context.Provider value={{ state, dispatch, saveFavoriteCard, removeFavoriteCard }}>
      {children}
    </Context.Provider>
  )
}
export default memo(ContextProvider)
