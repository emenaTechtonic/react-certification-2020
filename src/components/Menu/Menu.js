import React, { useContext } from 'react'
import { Link, useHistory } from 'react-router-dom'
import { Context } from '../../providers/Context/context'
import { AuthContext } from '../../providers/Auth/auth'
import StyledMenu from './Menu.styled'

const Menu = React.memo(() => {
  const { state, dispatch } = useContext(Context)
  const { authenticated } = useContext(AuthContext)
  const history = useHistory()

  const handleChange = (e) => {
    e.preventDefault()
    if (e.target.innerText === 'Favorites') {
      dispatch({ type: 'SET_VIEW', payload: 'favorites' })
      dispatch({ type: 'OPEN_MENU', payload: false })
      dispatch({ type: 'SET_TARGET', payload: state.target })
      history.push('/private')
    } else {
      dispatch({ type: 'SET_VIEW', payload: 'home' })
      dispatch({ type: 'OPEN_MENU', payload: false })
      dispatch({ type: 'SET_TARGET', payload: state.target })
      history.push('/')
    }
  }

  return (
    <StyledMenu toggle={state.menu}>
      <Link to="/" className="homeLink" onClick={handleChange}>
        Home
      </Link>
      {authenticated ? (
        <Link to="/private" className="favoritesLink" onClick={handleChange}>
          Favorites
        </Link>
      ) : (
        <div />
      )}
    </StyledMenu>
  )
})
export default Menu
