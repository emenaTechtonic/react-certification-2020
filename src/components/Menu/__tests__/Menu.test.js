import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent, waitFor } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { Context } from '../../../providers/Context/context'
import { AuthContext } from '../../../providers/Auth/auth'
import Menu from '../Menu'

const jsonFile = require('../../../utils/assets/youtube-videos-mock.json')

const login = null
const logout = null

let state = {
  target: 'Wizeline',
  theme: false,
  data: jsonFile,
  user: { id: 0, name: '', avatarUrl: '' },
  favorites: [],
}

let authenticated = false

describe('Testing Menu component', () => {
  test('Renders Menu without crashing', async () => {
    const rendered = renderer.create(
      <AuthContext.Provider value={{ authenticated, login, logout }}>
        <Context.Provider value={{ state }}>
          <MemoryRouter>
            <Menu />
          </MemoryRouter>
        </Context.Provider>
      </AuthContext.Provider>
    )
    expect(rendered).toMatchSnapshot()
  })

  test('Favorites view changed when button menu clicked', async () => {
    const e = { target: { innerText: 'Favorites' } }
    authenticated = true
    state = {
      view: 'home',
    }

    const payload = { payload: 'favorites', type: 'SET_VIEW' }

    const handleChange = jest.fn()
    const dispatch = jest.fn()

    const { getByText } = render(
      <Context.Provider value={{ state, dispatch }}>
        <MemoryRouter>
          <AuthContext.Provider value={{ authenticated, login, logout }}>
            <Menu handleChange={handleChange} />
          </AuthContext.Provider>
        </MemoryRouter>
      </Context.Provider>
    )

    const container = await waitFor(() => getByText('Favorites'))
    fireEvent.click(container, e)

    expect(dispatch).toHaveBeenCalledTimes(3)
    expect(dispatch).toHaveBeenCalledWith(payload)
  })

  test('Home view changed when button menu clicked', async () => {
    const e = { target: { innerText: 'Home' } }

    const payload = { payload: 'home', type: 'SET_VIEW' }

    const handleChange = jest.fn()
    const dispatch = jest.fn()

    const { getByText } = render(
      <AuthContext.Provider value={{ authenticated, login, logout }}>
        <Context.Provider value={{ state, dispatch }}>
          <MemoryRouter>
            <Menu handleChange={handleChange} />
          </MemoryRouter>
        </Context.Provider>
      </AuthContext.Provider>
    )

    const container = await waitFor(() => getByText('Home'))
    fireEvent.click(container, e)

    expect(dispatch).toHaveBeenCalledTimes(3)
    expect(dispatch).toHaveBeenCalledWith(payload)
  })
})
