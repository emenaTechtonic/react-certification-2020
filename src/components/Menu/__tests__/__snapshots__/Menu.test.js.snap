// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Menu component Renders Menu without crashing 1`] = `
<div
  className="sc-bdvvaa bYAoZE menu"
>
  <a
    className="homeLink"
    href="/"
    onClick={[Function]}
  >
    Home
  </a>
  <div />
</div>
`;
